src/role_management.c

```diff
@@ -0,0 +1,665 @@
+	if (rv != 0) {
+		return;
+	}
+	clientInit(&proto, sock);
+	rv = clientSendHandshake(&proto, &context);
+	if (rv != 0) {
+		goto close;
+	}
+	rv = clientSendDescribe(&proto, &context);
+	rv = clientRecvMetadata(&proto,
+			&polling->cluster[polling->i].failure_domain,
+			&polling->cluster[polling->i].weight, &context);
+	if (rv != 0) {
+		goto close;
+	}
+	polling->cluster[polling->i].online = true;
+
+close:
+	clientClose(&proto);
+}
+
+/* Runs on the main thread after polling each server for roles adjustment. */
+static void pollClusterAfterWorkCb(uv_work_t *work, int status)
+{
+	struct polling *polling = work->data;
+	uv_work_t *work_objs;
+	struct polling *polling_objs;
+	unsigned i;
+
+	if (status != 0) {
+		/* XXX */
+		return;
+	}
+	*polling->count += 1;
+	/* If all nodes have been polled, invoke the callback. */
+	if (*polling->count == polling->n_cluster) {
+		polling->cb(polling);
+		/* Free the shared data, now that all tasks have finished. */
+		raft_free(polling->count);
+		for (i = 0; i < polling->n_cluster; i += 1) {
+			raft_free(polling->cluster[i].address);
+		}
+		raft_free(polling->cluster);
+		work_objs = work - polling->i;
+		raft_free(work_objs);
+		polling_objs = polling - polling->i;
+		raft_free(polling_objs);
+	}
+}
+
+/* Poll every node in the cluster to learn whether it's online, and if so, its weight
+ * and failure domain. */
+static void pollCluster(struct dqlite_node *d, void (*cb)(struct polling *)) {
+	struct all_node_info *cluster;
+	const struct raft_server *server;
+	struct polling *polling_objs;
+	struct polling *polling;
+	struct uv_work_s *work_objs;
+	struct uv_work_s *work;
+	unsigned *count;
+	unsigned n;
+	unsigned i;
+	unsigned j;
+	unsigned ii;
+	int rv;
+
+	n = d->raft.configuration.n;
+	cluster = raft_calloc(n, sizeof *cluster);
+	if (cluster == NULL) {
+		goto err;
+	}
+	count = raft_malloc(sizeof *count);
+	if (count == NULL) {
+		goto err_after_alloc_cluster;
+	}
+	*count = 0;
+	for (i = 0; i < n; i += 1) {
+		server = &d->raft.configuration.servers[i];
+		cluster[i].id = server->id;
+		cluster[i].address = raft_malloc(strlen(server->address) + 1);
+		if (cluster[i].address == NULL) {
+			goto err_after_alloc_addrs;
+		}
+		memcpy(cluster[i].address, server->address, strlen(server->address) + 1);
+		cluster[i].role = translateRaftRole(server->role);
+	}
+	polling_objs = raft_calloc(n, sizeof *polling_objs);
+	if (polling_objs == NULL) {
+		goto err_after_alloc_addrs;
+	}
+	work_objs = raft_calloc(n, sizeof *work_objs);
+	if (work_objs == NULL) {
+		goto err_after_alloc_polling;
+	}
+	for (j = 0; j < n; j += 1) {
+		polling = &polling_objs[j];
+		polling->cb = cb;
+		polling->node = d;
+		polling->cluster = cluster;
+		polling->n_cluster = n;
+		polling->count = count;
+		polling->i = j;
+		work = &work_objs[j];
+		work->data = polling;
+		rv = uv_queue_work(&d->loop, work, pollClusterWorkCb, pollClusterAfterWorkCb);
```

Comment:
I think we might want to increase the libuv thread pool size, see https://docs.libuv.org/en/v1.x/threadpool.html something like 128 by default?

Reply:
Note that if the capacity of the thread pool gets exhausted, then further calls to `uv_queue_work` will not fail, but rather queued and executed when threads become available. In general the size of the threadpool should ideally match the hardware capacity (e.g. the number of cores of the host), in order to reduce contention and context switches.

I think we can safely leave libuv's default.